h4/h4.ino

```diff
@@ -2935,6 +2965,49 @@ bool handleMcode(const String& command) {
   return true;
 }
 
+bool handleChangeToolCommand(String command) {
+  int toolNumber = command.substring(1).toInt();
+  if (toolNumber >= 0 && toolNumber < 4) {
+    changeTool(toolNumber);
+    return true;
+  } else {
+    Serial.println("error: invalid tool number");
+    return false;
+  }
+}
+
+void changeTool(int newToolNumber) {
+    currentTool = newToolNumber;
+    ToolOffset newToolOffset = toolOffsets[newToolNumber];
+
+    // Calculate the net offset needed to apply
+    ToolOffset netOffset = {
+        newToolOffset.xOffset - currentOffset.xOffset,
+        newToolOffset.zOffset - currentOffset.zOffset
+    };
+
+    applyToolOffset(netOffset);
+    currentOffset = newToolOffset;
+
+    Serial.println("Tool changed to T" + String(newToolNumber));
+}
+
+void applyToolOffset(ToolOffset offset) {
+    long xOffsetSteps = duToSteps(&x, offset.xOffset);
+    long zOffsetSteps = duToSteps(&z, offset.zOffset);
+
+    // Lock the axes before making changes
```

Comment:
This has lead to deadlocks in the past, instead we should raise the flag that tool has changed that is checked from the main loop() - see https://github.com/kachurovskiy/nanoels/blob/main/h4/h4.ino#L3095C22-L3095C22 for example

Reply:
@kachurovskiy ive addressed the other comments but could you give a bit more info on this flag? Im not quite sure I follow how you would do it with flags.

Reply:
1. Create a new variable `int nextTool = 0`
2. Initialize it correctly on start
3. When a GCode `TX` command comes, set `nextTool` to `X`
4. In `void applySettings()` check if `tool != newTool` and if so, make the axis offset change by calling `applyTool()` which presumably would change axis position by difference between new and old tool offsets - though I'm not sure how exactly and which side effects that would have e.g. being suddenly located outside of limits